File: banking.py
"""
Bank Account System
Author: Arsh
Description: CLI-based Bank Account System demonstrating OOP concepts.
"""


class Account:
    """Demonstrates OOP — encapsulation, private attributes, and methods."""

    def __init__(self, account_no: int, full_name: str, balance: int, mobile_no: int, email: str, pin: int) -> None:
        self.account_no: int = account_no
        self.full_name: str = full_name
        self.mobile_no: int = mobile_no
        self.email: str = email

        self.__balance: int = balance
        self.__pin: int = pin

    def _get_valid_amount(self, prompt: str) -> int:
        """Loops until a valid positive integer amount is entered."""
        while True:
            try:
                amount = int(input(prompt))
                if amount <= 0:
                    print("Amount must be greater than 0.")
                    continue
                return amount
            except ValueError:
                print("Invalid input. Please enter a number.")

    def show_details(self) -> None:
        print("\n---- ACCOUNT DETAILS ----")
        print(f"{'Account No':<12} {self.account_no}")
        print(f"{'Full Name':<12} {self.full_name}")
        print(f"{'Mobile No':<12} {self.mobile_no}")
        print(f"{'Email':<12} {self.email}")

    def deposit(self) -> None:
        print("\n---- DEPOSIT ----")
        amount: int = self._get_valid_amount("Enter amount: ₹")
        self.__balance += amount
        print(f"Deposited ₹{amount} | Current Balance: ₹{self.__balance}")

    def withdraw(self) -> None:
        print("\n---- WITHDRAW ----")
        amount: int = self._get_valid_amount("Enter amount: ₹")
        if amount > self.__balance:
            print("Insufficient balance.")
            return
        self.__balance -= amount
        print(f"Withdrawn ₹{amount} | Balance Left: ₹{self.__balance}")

    def get_balance(self) -> int:
        return self.__balance

    def login(self) -> bool:
        """Allows 3 PIN attempts before locking out. Returns True on success."""
        attempts: int = 0

        while attempts < 3:
            print("\n---- LOGIN ----")
            try:
                pin: int = int(input("Enter PIN: "))
                if len(str(pin)) != 4:
                    print("PIN must be exactly 4 digits.")
                    attempts += 1
                    continue
                if pin == self.__pin:
                    print("Login successful.")
                    return True
                else:
                    print(f"Incorrect PIN. {2 - attempts} attempt(s) remaining.")
                    attempts += 1
            except ValueError:
                print("PIN must be digits only.")
                attempts += 1

        print("Too many failed attempts. Access denied.")
        return False


def show_menu() -> None:
    print("\n---- MAIN MENU ----")
    options: list[tuple[int, str]] = [
        (1, "Display Details"),
        (2, "Deposit"),
        (3, "Withdraw"),
        (4, "Check Balance"),
        (5, "Logout"),
    ]
    for option, action in options:
        print(f"{option}. {action}")


def ask_return_to_menu() -> bool:
    """Returns True to continue, False to exit."""
    while True:
        choice: str = input("\nReturn to Main Menu? (y/n): ").strip().lower()
        if choice in ("y", "yes"):
            return True
        elif choice in ("n", "no"):
            print("Goodbye! — Made by Arsh")
            return False
        else:
            print("Invalid input. Enter y or n.")


def main() -> None:
    account = Account(
        account_no=5001037423,
        full_name="Arshdeep Singh",
        balance=89000,
        mobile_no=9988776655,
        email="abc@xyz.com",
        pin=9833
    )

    if not account.login():
        return

    while True:
        show_menu()
        try:
            choice: int = int(input("Enter choice (1-5): "))

            if choice == 1:
                account.show_details()
            elif choice == 2:
                account.deposit()
            elif choice == 3:
                account.withdraw()
            elif choice == 4:
                print(f"\nCurrent Balance: ₹{account.get_balance()}")
            elif choice == 5:
                print("Thanks for using! — Made by Arsh")
                break
            else:
                print("Invalid option. Please choose between 1 and 5.")
                continue

            if not ask_return_to_menu():
                break

        except ValueError:
            print("Input must be a number.")


if __name__ == "__main__":
    main()

File: banking.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class AccountInterface {
public:
    virtual void deposit(double amount) = 0;       
    virtual void withdraw(double amount) = 0;      
    virtual void displayBalance() const = 0;       
    virtual ~AccountInterface() {}                 
};

class BankAccount : public AccountInterface {
private:
    string accountHolder;
    string pin;

protected:
    double balance; 

public:
    BankAccount(string name, string userPin, double initialBalance) {
        accountHolder = name;
        pin = userPin;
        balance = initialBalance;
    }

    bool validatePIN(const string& inputPin) const {
        return pin == inputPin;
    }

    string getAccountHolder() const {
        return accountHolder;
    }

    void deposit(double amount) override {
        if (amount > 0) {
            balance += amount;
            cout << "\n[Success] Successfully deposited ₹" << amount << endl;
        } else {
            cout << "\n[Error] Invalid deposit amount!" << endl;
        }
    }
};

class SavingsAccount : public BankAccount {
private:
    double interestRate;

public:
    SavingsAccount(string name, string userPin, double initialBalance, double rate)
        : BankAccount(name, userPin, initialBalance), interestRate(rate) {}

    void withdraw(double amount) override {
        if (amount <= 0) {
            cout << "\n[Error] Invalid withdrawal amount!" << endl;
        } else if (amount > balance) {
            cout << "\n[Error] Insufficient funds! Current balance: ₹" << balance << endl;
        } else {
            balance -= amount;
            cout << "\n[Success] Successfully withdrew ₹" << amount << endl;
        }
    }

    void displayBalance() const override {
        cout << "\n--- Savings Account Summary ---" << endl;
        cout << "Account Holder: " << getAccountHolder() << endl;
        cout << "Current Balance: ₹" << balance << endl;
        cout << "Interest Rate: " << interestRate << "% annually" << endl;
        cout << "-------------------------------" << endl;
    }
};

class CurrentAccount : public BankAccount {
private:
    double overdraftLimit;

public:
    CurrentAccount(string name, string userPin, double initialBalance, double limit)
        : BankAccount(name, userPin, initialBalance), overdraftLimit(limit) {}

    void withdraw(double amount) override {
        if (amount <= 0) {
            cout << "\n[Error] Invalid withdrawal amount!" << endl;
        } else if (amount > (balance + overdraftLimit)) {
            cout << "\n[Error] Overdraft limit exceeded! Max available: ₹" << (balance + overdraftLimit) << endl;
        } else {
            balance -= amount;
            cout << "\n[Success] Successfully withdrew ₹" << amount << endl;
        }
    }

    void displayBalance() const override {
        cout << "\n--- Current Account Summary ---" << endl;
        cout << "Account Holder: " << getAccountHolder() << endl;
        cout << "Current Balance: ₹" << balance << endl;
        cout << "Overdraft Limit: ₹" << overdraftLimit << endl;
        cout << "-------------------------------" << endl;
    }
};

int main() {
    AccountInterface* myAccount = nullptr;
    
    string inputPin;
    int trials = 3;
    bool isAuthenticated = false;

    cout << "========================================" << endl;
    cout << "      WELCOME TO ARSH BANKING SYSTEM    " << endl;
    cout << "========================================" << endl;

    int accType;
    cout << "\nSelect Account Type to Access:\n1. Savings Account\n2. Current Account\nChoice: ";
    cin >> accType;

    if (accType == 1) {
        myAccount = new SavingsAccount("Arshdeep Singh", "2508", 5000.0, 4.0);
    } else if (accType == 2) {
        myAccount = new CurrentAccount("Arshdeep Singh", "2508", 10000.0, 2000.0);
    } else {
        cout << "\n[Error] Invalid account type choice." << endl;
        return 0;
    }

    BankAccount* basePtr = dynamic_cast<BankAccount*>(myAccount);

    while (trials > 0) {
        cout << "\nPlease enter your 4-digit PIN (Attempts remaining: " << trials << "): ";
        cin >> inputPin;

        if (basePtr && basePtr->validatePIN(inputPin)) {
            isAuthenticated = true;
            break;
        } else {
            cout << "[Access Denied] Incorrect PIN structure or value." << endl;
            trials--;
        }
    }

    if (!isAuthenticated) {
        cout << "\n[ALERT] Too many failed attempts. Your account has been temporarily locked." << endl;
        delete myAccount;
        return 0; 
    }

    int choice;
    do {
        cout << "\n========= MAIN MENU =========" << endl;
        cout << "1. Check Balance" << endl;
        cout << "2. Deposit Funds" << endl;
        cout << "3. Withdraw Funds" << endl;
        cout << "4. Exit Application" << endl;
        cout << "Enter choice (1-4): ";
        cin >> choice;

        switch (choice) {
            case 1:
                myAccount->displayBalance();
                break;
            case 2: {
                double depAmount;
                cout << "Enter amount to deposit: ₹";
                cin >> depAmount;
                myAccount->deposit(depAmount);
                break;
            }
            case 3: {
                double witAmount;
                cout << "Enter amount to withdraw: ₹";
                cin >> witAmount;
                myAccount->withdraw(witAmount);
                break;
            }
            case 4:
                cout << "\nThank you for choosing Apex Banking. Goodbye!" << endl;
                break;
            default:
                cout << "\n[Error] Invalid choice! Please select an option between 1 and 4." << endl;
        }
    } while (choice != 4);

    delete myAccount;
    return 0;
}
